File: codes/rust/234.回文链表.rs
/*
 * @lc app=leetcode.cn id=234 lang=rust
 *
 * [234] 回文链表
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
        if head.is_none() {
            return true;
        }
        // 使用快慢指针找到链表中点
        let (mut fast, mut slow) = (head.as_ref(), head.as_ref());
        while fast.is_some() && fast.as_ref().unwrap().next.as_ref().is_some() {
            slow = slow.unwrap().next.as_ref();
            fast = fast.unwrap().next.as_ref().unwrap().next.as_ref();
        }
        // 反转后面的链表
        let mut prev = None;
        let mut curr = slow.cloned();
        while curr.is_some() {
            let next = curr.as_mut().unwrap().next.take();
            curr.as_mut().unwrap().next = prev.take();
            prev = curr;
            curr = next;
        }
        // 比较前后两个链表
        let mut left = head.as_ref();
        let mut right = prev.as_ref();
        while left.is_some() && right.is_some() {
            if left.as_ref().unwrap().val != right.as_ref().unwrap().val {
                return false;
            }
            left = left.unwrap().next.as_ref();
            right = right.unwrap().next.as_ref();
        }
        true
    }
}
// @lc code=end

File: codes/java/234.回文链表.java

class Solution {
	public boolean isPalindrome(ListNode head) {
		// 第一步: 使用快慢指针找到链表中点
		ListNode slow = head, fast = head;
		while (fast != null && fast.next != null) {
			slow = slow.next;
			fast = fast.next.next;
		}

		// 找到中点之后,反转后半部分链表，并对比前后两部分链表的值
		ListNode left = head, right = reverse(slow);
		while (left != null && right != null) {
			if (left.val != right.val) {
				return false;
			}
			left = left.next;
			right = right.next;
		}
		return true;
	}

	ListNode reverse(ListNode head) {
		ListNode prev = null, curr = head;
		while (curr != null) {
			ListNode next = curr.next;
			curr.next = prev;
			prev = curr;
			curr = next;
		}
		return prev;
	}
}